### PR TITLE
HPCC-23433 Fix locking race condition with RTM_DELETE_ON_DISCONNECT

### PR DESCRIPTION
**Pull request by jakesmith**
When deleting the root node on a RTM_DELETE_ON_DISCONNECT, a
mutex was being released prematurely, allowing another thread
to delete the root of the connection at the same time.
This in turn caused a 'Transaction to orphaned server node'
exception to be thrown.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
NB: a lot of the changes in this PR are cosmetic refactoring, resuling from the trivial CLCReadLockBlock, CLCWriteLockBlock classes being deleted in favour of their base class, so as to make it simpler to call the new changeToWrite method in the ReadWriteLock.

@mckellyln - please review.

File: system/jlib/jmutex.hpp [hunk 1]
```diff
@@ -586,67 +586,88 @@ class jlib_decl Monitor: public Mutex
...
+    {
+        cs.enter();
+        if (writeLocks)
+        {
+            cs.leave();
+            return false; // failed because already write locked
+        }
+        if (readLocks)
+            readLocks--;
+        return doLockWrite(timed, timeout); // returns false if timed out
```
Comment (mckellyln) [inline system/jlib/jmutex.hpp]: Should we try the doLockWrite() and then if ok decrement readLocks ?

File: system/jlib/jmutex.hpp [hunk 2]
```diff
@@ -586,67 +586,88 @@ class jlib_decl Monitor: public Mutex
...
+            readLocks++;
+            cs.leave();
+        }
+        else
+        {
+            readWaiting++;
+            cs.leave();
+            if (timed)
+            {
+                if (!readSem.wait(timeout)) {
```
Comment (mckellyln) [inline system/jlib/jmutex.hpp]: Might as well move these parens as well

**Comment by mckellyln** [comment]
@jakesmith one question and one comment.

File: system/jlib/jmutex.hpp (hunk 2 above)
Comment (jakesmith) [reply system/jlib/jmutex.hpp]: yep agree, will change formatting here, and below.

File: system/jlib/jmutex.hpp (hunk 1 above)
Comment (jakesmith) [reply system/jlib/jmutex.hpp]: yes, but in fact that introduces a whole in itself. It should not release/reduce the readLocskc count here 1st.
Unfortunately, that makes the implementation trickier in that there may be something in a new waiting state.
I'll revisit.

**Comment by jakesmith** [comment]
@ghalliday @mckellyln - more complicated than I had initially anticipated (in ReadWriteLock)

Please review.

File: system/jlib/jmutex.hpp [hunk 3]
```diff
@@ -661,35 +710,74 @@ class jlib_decl ReadWriteLock
...
-                                cs.leave();
-                            }
+        exclWriteOwner = 0;
+#endif
+        return true;
+    }
+    void unlock()
+    {
+        cs.enter(); 
+        if (readLocks)
```
Comment (ghalliday) [inline system/jlib/jmutex.hpp]: Could do with some comments.  E.g. if (readLocks is non zero then a read lock must be being released)

File: system/jlib/jmutex.hpp [hunk 4]
```diff
@@ -661,35 +710,74 @@ class jlib_decl ReadWriteLock
...
+        changeToWrite(false, 0);
+    }
+    bool changeToWrite(unsigned timeout)
+    {
+        return changeToWrite(true, timeout);
+    }
+    bool changeToRead()
+    {
+        cs.enter();
+        if (!writeLocks)
```
Comment (ghalliday) [inline system/jlib/jmutex.hpp]: minor - this could be checked outside the critical section.  Should it be an internal error rather than returning false?

File: system/jlib/jmutex.hpp [hunk 5]
```diff
@@ -586,71 +586,120 @@ class jlib_decl Monitor: public Mutex
...
+        }
+#ifdef _DEBUG
+        exclWriteOwner = GetCurrentThreadId();
+#endif
+        return true;
+    }
+    bool changeToWrite(bool timed, unsigned timeout)
+    {
+        cs.enter();
+        if (writeLocks || (0 == readLocks))
```
Comment (ghalliday) [inline system/jlib/jmutex.hpp]: more: should this  throw an internal error instead of returning false?  ((readLocks==0) could be checked before the cs.

Comment (ghalliday) [reply system/jlib/jmutex.hpp]: or is the intention that it is ok to call this if it already has a write lock.  In which case the comment on the return is misleading.

**Review by ghalliday — COMMENTED** [review]
@jakesmith I have looked at it and the change looks good.  (Although the code in jlib could do with more comments, and I'm not convinced it is good for it to all be inline).
@mckellyln I think it is worth you checking carefully as well.

File: system/jlib/jmutex.hpp (hunk 5 above)
Comment (jakesmith) [reply system/jlib/jmutex.hpp]: it's supposed to fail if already has a write lock, but a soft failure (return false) as opposed to an exception - in keeping with pre-exisitng timed lock read / write calls that return false on failure to get lock.

Comment (jakesmith) [reply system/jlib/jmutex.hpp]: >((readLocks==0) could be checked before the cs.

should it though ? it's an edge case, not generally expected, and I think code clearer when checked inside the crit.

File: system/jlib/jmutex.hpp (hunk 4 above)
Comment (jakesmith) [reply system/jlib/jmutex.hpp]: I think a boolean result indicating not locked failure, similar to the failure on timed lock read/write calls.

>this could be checked outside the critical section

I'm not sure this edge case makes it worth moving the test, and clearer semantics if inside crit ?

File: system/jlib/jmutex.hpp (hunk 3 above)
Comment (jakesmith) [reply system/jlib/jmutex.hpp]: will add some comments.

**Comment by jakesmith** [comment]
@ghalliday - thanks, added some replies and questions back re. whether some of these conditions really should be moved before locking the crit.

**Comment by jakesmith** [comment]
.. and @mckellyln - please also review.

File: system/jlib/jmutex.hpp (hunk 4 above)
Comment (ghalliday) [reply system/jlib/jmutex.hpp]: I don't really mind if it is inside or outside.
If the function should never be called in that situation then it should be an internal error.  If it is expected then there should be a comment e.g.
//If the lock is already a read lock then this function will return false.

Similarly for changeToWrite - if the lock is already a write lock then it will return false.
Is that the correct response though?  If there is a timeout you can't distinguish between timedout and couldn't get the lock, and already had the write lock.

I don't have a strong view on which semantics you choose, but it should have a comment documenting what they are.

File: system/jlib/jmutex.hpp (hunk 5 above)
Comment (ghalliday) [reply system/jlib/jmutex.hpp]: The comment implies it is not legal to call it when not read locked.  If that is kind-of allowed the comment should be changed.  E.g.
//if called when not readlocked then fail immediately.

File: system/jlib/jmutex.hpp (hunk 4 above)
Comment (jakesmith) [reply system/jlib/jmutex.hpp]: >If there is a timeout you can't distinguish between timedout and couldn't get the lock, and already had the write lock.

ok yes I agree, on reflection I think these unexpected use cases should throw an exception instead. Will change.

**Comment by jakesmith** [comment]
@mckellyln @ghalliday - changed invalid use cases to exceptions and added comments.

Have also moved some inline methods out of header into cpp, partly because use of jexcept inside header was causing circular header dependency issues.

**Comment by mckellyln** [comment]
@jakesmith ok, reviewiing still

File: system/jlib/jmutex.cpp [hunk 6]
```diff
@@ -321,7 +321,191 @@ void Monitor::notifyAll()
...
+            writeWaiting--;
+            writeLocks++;
+            writeSem.signal();
+        }
+    }
+    else // implies this unlock() is paired with a previous write lock
+    {
+        writeLocks--;
+        dbgassertex(writeLocks == 0);
+        if (readWaiting)
```
Comment (mckellyln) [inline system/jlib/jmutex.cpp]: So we are ok with readers always acquiring before any writers ?
Perhaps the check for writers should come first ?  Or we could rotate the order of this check on each unlock ?

Comment (jakesmith) [reply system/jlib/jmutex.cpp]: not sure, but that is not new semantics, ReadWriteLock has always given priority to pending blocked readers over blocked writers.

Worth considering, but as a separate task/discussion/JIRA.

**Comment by jakesmith** [comment]
>@jakesmith ok, reviewiing still

@mckellyln - re-tagging

**Review by mckellyln — APPROVED** [review]
I am ok with this, but I do think writers should get priority over readers.

**Comment by mckellyln** [comment]
@jakesmith ok for merge.

**Comment by jakesmith** [comment]
>I am ok with this, but I do think writers should get priority over readers.

as I say, worth considering, but wouldn't want change the pre-existing semantics in this PR.
Please open new JIRA to look at / further discussion.

**Comment by jakesmith** [comment]
@ghalliday - will squash ready for merge.

**Comment by jakesmith** [comment]
@ghalliday - please merge.

File: system/jlib/jmutex.cpp [hunk 7]
```diff
@@ -321,7 +321,191 @@ void Monitor::notifyAll()
...
+#ifdef _DEBUG
+    exclWriteOwner = GetCurrentThreadId();
+#endif
+    return true;
+}
+
+bool ReadWriteLock::changeToRead()
+{
+    cs.enter();
+    if (!writeLocks) // NB: If the lock is already a read lock then this function will return false.
```
Comment (ghalliday) [inline system/jlib/jmutex.cpp]: I'm not sure the comment is correct

**Review by ghalliday — COMMENTED** [review]
@jakesmith one minor comment on a comment

File: system/jlib/jmutex.cpp (hunk 7 above)
Comment (jakesmith) [reply system/jlib/jmutex.cpp]: no, should have been removed.

**Comment by jakesmith** [comment]
@ghalliday - removed rogue comment.
